lib/cretonne/src/verifier.rs

```diff
@@ -147,12 +147,109 @@ impl<'a> Verifier<'a> {
 
     fn instruction_integrity(&self, inst: Inst) -> Result<()> {
         let inst_data = &self.func.dfg[inst];
+        let dfg = &self.func.dfg;
 
         // The instruction format matches the opcode
         if inst_data.opcode().format() != InstructionFormat::from(inst_data) {
             return err!(inst, "instruction opcode doesn't match instruction format");
         }
 
+        let fixed_results = inst_data.opcode().constraints().fixed_results();
+        // var_results is 0 if we aren't a call instruction
+        let var_results = dfg.call_signature(inst)
+            .map(|sig| dfg.signatures[sig].return_types.len())
+            .unwrap_or(0);
+        let total_results = fixed_results + var_results;
+
+        if total_results == 0 {
+            // Instructions with no results have a NULL `first_type()`
+            let ret_type = inst_data.first_type();
+            if ret_type != types::VOID {
+                return err!(inst,
+                            "instruction expected to have NULL return value, found {}",
+                            ret_type);
+            }
+        } else {
+            // All result values for multi-valued instructions are created
+            let got_results = dfg.inst_results(inst).count();
+            if got_results != total_results {
+                return err!(inst,
+                            "expected {} result values, found {}",
+                            total_results,
+                            got_results);
+            }
+        }
+
+        self.verify_entity_references(inst)
+    }
+
+    fn verify_entity_references(&self, inst: Inst) -> Result<()> {
+        macro_rules! verify_ebb {
+            ($e:expr) => {
+                if !self.func.dfg.ebbs.is_valid(e) {
+                    return err!(inst, "invalid ebb reference");
+                }
+            };
+        }
+        macro_rules! verify_sig_ref {
+            ($s:expr) => {
+                if !self.func.dfg.signatures.is_valid(s) {
+                    return err!(inst, "invalid signature reference");
+                }
+            };
+        }
+        macro_rules! verify_func_ref {
+            ($f:expr) => {
+                if !self.func.dfg.ext_funcs.is_valid(f) {
+                    return err!(inst, "invalid function reference");
+                }
+            };
+        }
+        macro_rules! verify_value_list {
+            ($l:expr) => {
+                if self.func.dfg.value_lists.len_of(l) == None {
```

Comment:
Huh. The `len_of()` method wasn't supposed to be accessible outside the `entity_list` module. I think I would prefer if we could add some kind of `is_valid()` method instead.